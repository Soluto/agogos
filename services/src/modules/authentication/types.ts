import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
