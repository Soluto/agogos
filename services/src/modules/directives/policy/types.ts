import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
