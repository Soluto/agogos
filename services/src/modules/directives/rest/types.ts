export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
