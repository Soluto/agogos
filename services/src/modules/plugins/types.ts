import { BaseSchema } from '../base-schema';
import { ResourceGroup } from '../resource-repository';

export type ValueOrPromise<T> = T | Promise<T>;

export interface StitchPlugin {
  name: string;

  addArgumentInjectionGlobals?(): ValueOrPromise<Record<string, unknown>>;

  transformResourceGroup?(resourceGroup: ResourceGroup): ValueOrPromise<ResourceGroup>;

  transformBaseSchema?(baseSchema: BaseSchema): ValueOrPromise<BaseSchema>;
}
