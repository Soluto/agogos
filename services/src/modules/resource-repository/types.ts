import { RemoteSchema } from '../directives/gql';
import { Policy, LoadedPolicy } from '../directives/policy/types';

export type PolicyAttachments = Record<string, LoadedPolicy>;
export type PolicyQueryVariables = Record<string, unknown>;
export type PolicyArgsObject = Record<string, unknown>;

export interface ResourceGroup {
  schemas: Schema[];
  upstreams: Upstream[];
  upstreamClientCredentials: UpstreamClientCredentials[];
  policies: PolicyDefinition[];
  // policyAttachments are compiled from the Rego code in opa policies, they are not directly modified by users
  policyAttachments?: PolicyAttachments;
  basePolicy?: Policy;
  remoteSchemas?: RemoteSchema[];
}

export interface Resource {
  metadata: ResourceMetadata;
}

export interface ResourceMetadata {
  namespace: string;
  name: string;
}

export interface Schema extends Resource {
  schema: string;
}

export interface Upstream extends Resource {
  host: string;
  auth: {
    type: AuthType;
    activeDirectory: {
      authority: string;
      resource: string;
    };
  };
}

export interface UpstreamClientCredentials extends Resource {
  authType: AuthType;
  activeDirectory: {
    authority: string;
    clientId: string;
    clientSecret: string;
  };
}

export interface PolicyDefinition extends Resource {
  type: PolicyType;
  code: string;
  args?: PolicyArgsObject;
  query?: PolicyQuery;
  shouldOverrideBasePolicy?: boolean;
}

export interface PolicyQuery {
  gql: string;
  variables?: PolicyQueryVariables;
}

export interface FetchLatestResult {
  isNew: boolean;
  resourceGroup: ResourceGroup;
}

export type UpdateOptions = { registry: boolean };

export interface IResourceRepository {
  fetchLatest(): Promise<FetchLatestResult>;
  update(rg: ResourceGroup, options?: UpdateOptions): Promise<void>;
  writePolicyAttachment(filename: string, content: Buffer): Promise<void>;
}

enum AuthType {
  ActiveDirectory = 'ActiveDirectory',
}

export enum PolicyType {
  opa = 'opa',
}
