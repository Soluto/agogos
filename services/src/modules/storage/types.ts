export interface Storage {
  fileStats(filePath: string): Promise<FileStats>;
  writeFile(filePath: string, content: string | Buffer): Promise<void>;
  listFiles(folderPath: string): Promise<listFilesItem[]>;

  readFile(filePath: string): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: false }): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  readFile(filePath: string, options?: { asString?: boolean }): Promise<{ content: Buffer | string }>;
}

export type FileStats = {
  lastModified: Date;
};

export type listFilesItem = {
  filePath: string;
  lastModified: Date;
};
