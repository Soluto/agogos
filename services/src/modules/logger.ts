import * as pino from 'pino';
import { logLevel, nodeEnv, loggerConfiguration } from './config';

const loggerConfig: pino.LoggerOptions = {
  level: logLevel.toLowerCase(),
  timestamp: pino.stdTimeFunctions.unixTime,
  messageKey: 'message',
};

const devLoggerConfig: pino.LoggerOptions = {
  level: 'trace',
  prettyPrint: {
    colorize: true,
    translateTime: 'HH:MM:ss.l',
    ignore: 'pid,hostname',
  },
};

export default pino({
  ...loggerConfig,
  ...(nodeEnv !== 'production' ? devLoggerConfig : {}),
  ...loggerConfiguration,
});
