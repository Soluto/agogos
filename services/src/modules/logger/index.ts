import rootLogger from './root-logger';
import createChildLogger from './child-logger';

export { createChildLogger };
export default rootLogger;
