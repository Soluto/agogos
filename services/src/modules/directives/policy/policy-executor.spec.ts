import { GraphQLResolveInfo } from 'graphql';
import { when } from 'jest-when';
import logger from '../../logger';
import {
  PolicyArgsDefinitions,
  PolicyArgsObject,
  PolicyDefinition,
  PolicyType,
  ResourceMetadata,
} from '../../resource-repository';
import { Policy, PolicyEvaluationContext } from './types';

interface TestCase {
  policyArgs?: PolicyArgsObject;
  policyArgsDefinitions?: PolicyArgsDefinitions;
  expectedArgs?: Record<string, unknown>;
  shouldThrow?: boolean;
}

const testCases: [string, TestCase][] = [
  [
    'Without args',
    {
      expectedArgs: {},
    },
  ],
  [
    'Empty args',
    {
      policyArgs: {},
      policyArgsDefinitions: {},
      expectedArgs: {},
    },
  ],
  [
    'One required arg. Has value',
    {
      policyArgs: { arg1: 'Hello' },
      policyArgsDefinitions: { arg1: { type: 'String!' } },
      expectedArgs: { arg1: 'Hello' },
    },
  ],
  [
    'One required arg. Has no value',
    {
      policyArgs: {},
      policyArgsDefinitions: { arg1: { type: 'String!' } },
      shouldThrow: true,
    },
  ],
  [
    'One required arg. Has null value',
    {
      policyArgs: { arg1: null },
      policyArgsDefinitions: { arg1: { type: 'String!' } },
      expectedArgs: { arg1: null },
    },
  ],
  [
    'One required arg. Has undefined value',
    {
      policyArgs: { arg1: undefined },
      policyArgsDefinitions: { arg1: { type: 'String!' } },
      shouldThrow: true,
    },
  ],
  [
    'One optional arg. Has no value',
    {
      policyArgs: {},
      policyArgsDefinitions: { arg1: { type: 'String', optional: true } },
      expectedArgs: { arg1: null },
    },
  ],
  [
    'One optional arg. Has no value and default',
    {
      policyArgs: {},
      policyArgsDefinitions: { arg1: { type: 'String', default: 'Hey', optional: true } },
      expectedArgs: { arg1: 'Hey' },
    },
  ],
  [
    'One required arg. Has no value and default',
    {
      policyArgs: {},
      policyArgsDefinitions: { arg1: { type: 'String!', default: 'Hey' } },
      expectedArgs: { arg1: 'Hey' },
    },
  ],
  [
    'One required arg. Has value and default',
    {
      policyArgs: { arg1: 'Hello' },
      policyArgsDefinitions: { arg1: { type: 'String!', default: 'Hey' } },
      expectedArgs: { arg1: 'Hello' },
    },
  ],
  [
    'Two arguments',
    {
      policyArgs: { arg1: 'Hello' },
      policyArgsDefinitions: { arg1: { type: 'String!' }, arg2: { type: 'String!', default: 'World' } },
      expectedArgs: { arg1: 'Hello', arg2: 'World' },
    },
  ],
  [
    'Two arguments with injected values',
    {
      policyArgs: { arg1: '{ "H" + "I" }' },
      policyArgsDefinitions: { arg1: { type: 'String!' }, arg2: { type: 'String!', default: 'W{1-1}rld' } },
      expectedArgs: { arg1: 'HI', arg2: 'W0rld' },
    },
  ],
];

const typedUndefined = <T>() => (undefined as unknown) as T;

describe('Policy Executor', () => {
  const policyMetadata: ResourceMetadata = {
    namespace: 'policy-executor-test',
    name: 'policy',
  };

  const opaEvaluateMock = jest.fn();

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it.each(testCases)('%s', async (_, { policyArgs, policyArgsDefinitions, expectedArgs, shouldThrow }) => {
    const ctx: PolicyEvaluationContext = {
      ...policyMetadata,
      args: expectedArgs,
      query: undefined,
      policyAttachments: typedUndefined(),
    };

    when(opaEvaluateMock)
      .mockReturnValue({ done: true, allow: false })
      .calledWith(ctx)
      .mockReturnValue({ done: true, allow: true });
    jest.mock('./opa', () => ({
      evaluate: opaEvaluateMock,
    }));

    const { default: PolicyExecutor } = await import('./policy-executor');

    const executor = new PolicyExecutor();

    const policy: Policy = {
      ...policyMetadata,
      args: policyArgs,
    };

    const policyDefinition: PolicyDefinition = {
      metadata: policyMetadata,
      type: PolicyType.opa,
      code: 'Some code',
      args: policyArgsDefinitions,
    };

    const context = {
      resourceGroup: {
        policies: [policyDefinition],
      },
    } as any;

    const info = ({
      parentType: {
        name: 'Foo',
      },
      fieldName: 'bar',
    } as unknown) as GraphQLResolveInfo;

    try {
      const result = await executor.evaluatePolicy(policy, typedUndefined(), typedUndefined(), context, info, logger);
      expect(result).toBeTruthy();
      expect(shouldThrow).not.toBeTruthy();
    } catch (err) {
      expect(shouldThrow).toBeTruthy();
      expect(err).toMatchSnapshot();
    }
  });
});
