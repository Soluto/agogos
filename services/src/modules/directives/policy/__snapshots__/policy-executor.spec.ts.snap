// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Executor One required arg. Has no value 1`] = `[PolicyExecutionFailedError: Policy "policy" in namespace "policy-executor-test" execution failed: Missing argument "arg1"]`;

exports[`Policy Executor One required arg. Has null value 1`] = `[PolicyExecutionFailedError: Policy "policy" in namespace "policy-executor-test" execution failed: Non-nullable argument "arg1" got value "null"]`;

exports[`Policy Executor One required arg. Has undefined value 1`] = `[PolicyExecutionFailedError: Policy "policy" in namespace "policy-executor-test" execution failed: Missing argument "arg1"]`;
